### ISSUE TITLE
mistruths in perlunicode.pod

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1452]
Migrated from [rt.perl.org#2644](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2644) (status was 'resolved')

Searchable as RT2644$

**Comment by p5pRT** [comment]
### From root@dixie.cscaper.com

#### Created by joseph@5sigma.com



&nbsp;       o   Strings and patterns may contain characters that have
&nbsp;           an ordinal value larger than 255\.  In Perl v5\.6\, this
&nbsp;           is only enabled if the lexical scope has a use utf8
&nbsp;           declaration \(due to compatibility needs\) but future
&nbsp;           versions may enable this by default\.

This is apparently untrue\.

&nbsp;  print ord\(v2000\)\, "\\n";
&nbsp;  print ord\("\\x\{7d0\}"\)\, "\\n";

Both print "2000" regardless of whether use utf8 is in effect\.  

I'm really not sure HOW this is supposed to work\, but if this is correct
behavior\, it seems like the explanation would have to refer to the fact
that there is a bit that indicates whether strings are byte\- or UTF\-8\-
encoded and that\, no matter whether use utf8 is in effect\, there are
constructs that generate UTF\-8\-encoded strings\.  If this is undesirable
\(personally\, I think being able to generate UTF\-8\-encoded strings
without use utf8 \*is\* desirable\) then the holes that allow the
generation of UTF\-8\-encoded strings in this version of perl without 
use utf8 should be closed\.

Also\, it seems like there ought to be a built\-in function \(available
WITHOUT use utf8\) that can be used to determine whether a string is
UTF\-8\-encoded\.  Or at least it should be in a separate pragma module
or available in a way that doesn't inflict the semantics of utf8 on
programs\.  Something like isutf8\($str\) = TRUE if $str is UTF\-8\-encoded\,
or perhaps encoding\($str\) = 'UTF\-8'\.  Maybe there is such a built\-in
or pragma already but I haven't found it in the docs\.  \(Yet\.\)

<details><summary>Perl Info</summary>

```


Site configuration information for perl v5.5.670:

Configured by joseph at Wed Mar 22 17:29:59 MST 2000.

Summary of my perl5 (revision 5.0 version 5 subversion 670) configuration:
  Platform:
    osname=linux, osvers=2.2.12-20, archname=i586-linux-64all
    uname='linux dixie.cscaper.com 2.2.12-20 #1 mon sep 27 10:25:54 edt 1999 i586 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=define use64bitall=define uselongdouble=define usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.5.670:
    /usr/local/devel/perl-longlong/lib/5.5.670/i586-linux-64all
    /usr/local/devel/perl-longlong/lib/5.5.670
    /usr/local/devel/perl-longlong/lib/site_perl/5.5.670/i586-linux-64all
    /usr/local/devel/perl-longlong/lib/site_perl/5.5.670
    /usr/local/devel/perl-longlong/lib/site_perl
    .


Environment for perl v5.5.670:
    HOME=/root
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/sbin:/sbin:/bin:/usr/bin:/home/joseph/bin:/usr/local/bin:/usr/X11R6/bin:/usr/local/frame/bin:/usr/local/mysql/bin:/oracle/sw/bin:/usr/sbin:/usr/games:.
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 23 Mar 2000 22&#8203;:02&#8203;:10 MST\, root wrote&#8203;:
>
>This is a bug report for perl from joseph@&#8203;5sigma\.com\,
>generated with the help of perlbug 1\.27 running under perl v5\.5\.670\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>       o   Strings and patterns may contain characters that have
>           an ordinal value larger than 255\.  In Perl v5\.6\, this
>           is only enabled if the lexical scope has a use utf8
>           declaration \(due to compatibility needs\) but future
>           versions may enable this by default\.

Lose 5\.5\.670\, get 5\.6\.0\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, 23 Mar 2000 21&#8203;:18&#8203;:34 \-0800 gsar@&#8203;ActiveState\.com \(Gurusamy Sarathy\)
wrote&#8203;:

\* On Thu\, 23 Mar 2000 22&#8203;:02&#8203;:10 MST\, root wrote&#8203;:
\* >
\* >This is a bug report for perl from joseph@&#8203;5sigma\.com\,
\* >generated with the help of perlbug 1\.27 running under perl v5\.5\.670\.
\* >
\* >
\* >\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\* >\[Please enter your report here\]
\* >
\* >       o   Strings and patterns may contain characters that have
\* >           an ordinal value larger than 255\.  In Perl v5\.6\, this
\* >           is only enabled if the lexical scope has a use utf8
\* >           declaration \(due to compatibility needs\) but future
\* >           versions may enable this by default\.
\*
\* Lose 5\.5\.670\, get 5\.6\.0\.

Okay\, I am looking over these issues with 5\.6\.0 and I still see \.\.\.

&nbsp;  "\\x\{c8\}" still equivalent to "\\xc8"&#8203;: 1 byte w/o utf8\, 2 bytes with

The following prints 2000 twice\, regardless of use utf8 \(same as before\)

&nbsp;  print ord\(v2000\)\, "\\n";
&nbsp;  print ord\("\\x\{7d0\}"\)\, "\\n";

use utf8;
print "\\\\x\{41\} and v65 are different\\n" unless "\\x\{41\}" eq \+v65;
print "\\\\x41 and v65 are different\\n" unless "\\x41" eq \+v65;
print "\\\\x\{c8\} and v200 are different\\n" unless "\\x\{c8\}" eq \+v200;
print "\\\\xc8 and v200 are different\\n" unless "\\xc8" eq \+v200;
print "\\\\x\{221b\} and v8731 are different\\n" unless "\\x\{221b\}" eq v8731;

still yields

\\x\{c8\} and v200 are different
\\xc8 and v200 are different

print v200;
print eval "v200";
print eval "\+v200";

still prints

??v200??

printf "%d\\n"\, 3000000000;
printf "%lld\\n"\, 3000000000;

still prints

\-1294967296
3000000000

on a long long build\.

In other words\, all the stuff I just perlbugged is apparently still 
the same\, as is some other stuff that was just too messy to report\, 
although there appear to be some documentation changes\.

&nbsp;  \-joseph

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 23 Mar 2000 23&#8203;:30&#8203;:48 MST\, "Joseph N\. Hall" wrote&#8203;:
>Okay\, I am looking over these issues with 5\.6\.0 and I still see \.\.\.
>
>  "\\x\{c8\}" still equivalent to "\\xc8"&#8203;: 1 byte w/o utf8\, 2 bytes with

I think \\xc8 being one or two bytes was intentional \(to preserve
compatibility\)\.  But "\\x\{NN\}" behaving the same way as "\\xNN"
\(for 127 \< NN \< 256\) could be construed a misfeature\, because there's
no compatibility issue to worry about with \\x\{\.\.\.\}\.

>The following prints 2000 twice\, regardless of use utf8 \(same as before\)
>
>  print ord\(v2000\)\, "\\n";
>  print ord\("\\x\{7d0\}"\)\, "\\n";

It's not clear why you think this is a bug\.  Seems correct to me\.

>use utf8;
>print "\\\\x\{41\} and v65 are different\\n" unless "\\x\{41\}" eq \+v65;
>print "\\\\x41 and v65 are different\\n" unless "\\x41" eq \+v65;
>print "\\\\x\{c8\} and v200 are different\\n" unless "\\x\{c8\}" eq \+v200;
>print "\\\\xc8 and v200 are different\\n" unless "\\xc8" eq \+v200;
>print "\\\\x\{221b\} and v8731 are different\\n" unless "\\x\{221b\}" eq v8731;
>
>still yields
>
>\\x\{c8\} and v200 are different
>\\xc8 and v200 are different

This is the same issue as the first one\.

>print v200;
>print eval "v200";
>print eval "\+v200";
>
>still prints
>
>Èv200È

I think Perl should be fixed to allow vNNN when PL\_expect == XSTATE\.
\(see toke\.c\, line\#3539\)\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri\, 24 Mar 2000 07&#8203;:05&#8203;:58 \-0800 gsar@&#8203;ActiveState\.com \(Gurusamy Sarathy\)
wrote&#8203;:

\* On Thu\, 23 Mar 2000 23&#8203;:30&#8203;:48 MST\, "Joseph N\. Hall" wrote&#8203;:
\* >Okay\, I am looking over these issues with 5\.6\.0 and I still see \.\.\.
\* >
\* >  "\\x\{c8\}" still equivalent to "\\xc8"&#8203;: 1 byte w/o utf8\, 2 bytes with
\*
\* I think \\xc8 being one or two bytes was intentional \(to preserve
\* compatibility\)\.  But "\\x\{NN\}" behaving the same way as "\\xNN"
\* \(for 127 \< NN \< 256\) could be construed a misfeature\, because there's
\* no compatibility issue to worry about with \\x\{\.\.\.\}\.

It's at odds with the documentation\, which says

&nbsp;           A character in the Latin\-1 range
&nbsp;           \(128\.\.255\) should be written \\x\{ab\} rather than \\xab\,
&nbsp;           since the former will turn into a two\-byte UTF\-8 code\,
&nbsp;           while the latter will continue to be interpreted as
&nbsp;           generating a 8\-bit byte rather than a character\.

\*
\* >The following prints 2000 twice\, regardless of use utf8 \(same as before\)
\* >
\* >  print ord\(v2000\)\, "\\n";
\* >  print ord\("\\x\{7d0\}"\)\, "\\n";
\*
\* It's not clear why you think this is a bug\.  Seems correct to me\.

This also WAS at odds with the documentation on use utf8\, which said that
strings could only have wide characters under use utf9\.  But that has
been changed\.  So\, I have no particular objection if they're supposed to
work that way\.

\* This is the same issue as the first one\.
\*
\* >print v200;
\* >print eval "v200";
\* >print eval "\+v200";
\* >
\* >still prints
\* >
\* >??v200??
\*
\* I think Perl should be fixed to allow vNNN when PL\_expect == XSTATE\.
\* \(see toke\.c\, line\#3539\)\.

I think v200 should NOT work in any case\.  :\-\)  It's not too late to stop
this particular bad thing\.  Please\, consider not adding icky things like
this that don't do anything really useful\.  What's wrong with v\{200\}
instead?

The one other bad thing is does vnnn\.nnn have character or byte semantics?
I think you need to pick one\, and I think it has to be character\.  As
cool as vnnn\.nnn is for constructing byte strings\, I can't see any way
that it should work that way unless you do&#8203;:

&nbsp;  \{ use bytes; $byte\_str = v200\.200\.200 \}

\-\-assuming that utf8 is intended to be mostly a no\-op in the near future\.

The current "smarts" that make v200\.200 bytes and v200\.2000 characters
are Bad Bad Bad\.  I hope you can see this\.  What if I want v200\.200 
to be characters? 

&nbsp;  \-joseph

**Comment by p5pRT** [comment]
### From @gsar

On Fri\, 24 Mar 2000 10&#8203;:28&#8203;:53 MST\, "Joseph N\. Hall" wrote&#8203;:
>On Fri\, 24 Mar 2000 07&#8203;:05&#8203;:58 \-0800 gsar@&#8203;ActiveState\.com \(Gurusamy Sarathy\)
>wrote&#8203;:
>\* I think \\xc8 being one or two bytes was intentional \(to preserve
>\* compatibility\)\.  But "\\x\{NN\}" behaving the same way as "\\xNN"
>\* \(for 127 \< NN \< 256\) could be construed a misfeature\, because there's
>\* no compatibility issue to worry about with \\x\{\.\.\.\}\.
>
>It's at odds with the documentation\, which says
>
>           A character in the Latin\-1 range
>           \(128\.\.255\) should be written \\x\{ab\} rather than \\xab\,
>           since the former will turn into a two\-byte UTF\-8 code\,
>           while the latter will continue to be interpreted as
>           generating a 8\-bit byte rather than a character\.

Yes\, I think the documentation needs fixing here\.

>\* >  print ord\(v2000\)\, "\\n";
>\* >  print ord\("\\x\{7d0\}"\)\, "\\n";
>\*
>\* It's not clear why you think this is a bug\.  Seems correct to me\.
>
>This also WAS at odds with the documentation on use utf8\, \[\.\.\.\]

You forget that developer versions have a right to be "odd"\.  :\-\)

>\* I think Perl should be fixed to allow vNNN when PL\_expect == XSTATE\.
>\* \(see toke\.c\, line\#3539\)\.
>
>I think v200 should NOT work in any case\.  :\-\)  It's not too late to stop
>this particular bad thing\.  Please\, consider not adding icky things like
>this that don't do anything really useful\.  What's wrong with v\{200\}
>instead?

My original implementation didn't allow vNNN\, but it was more surprising
to people that it didn't work\, so I changed my mind\.

I obviously disagree with you on this\.

>The current "smarts" that make v200\.200 bytes and v200\.2000 characters
>are Bad Bad Bad\.  I hope you can see this\.  What if I want v200\.200 
>to be characters? 

Er\, v200\.200 is \_always\_ characters\.  What makes you think it isn't?


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\* >
\* >I think v200 should NOT work in any case\.  :\-\)  It's not too late to stop
\* >this particular bad thing\.  Please\, consider not adding icky things like
\* >this that don't do anything really useful\.  What's wrong with v\{200\}
\* >instead?
\*
\* My original implementation didn't allow vNNN\, but it was more surprising
\* to people that it didn't work\, so I changed my mind\.
\*
\* I obviously disagree with you on this\.

Well \.\.\. you have&#8203;:

&nbsp;  vnnn
&nbsp;  void vnnn
&nbsp;  \+vnnn
&nbsp;  \-vnnn

and all of the above with/without sub vnnn and/or other bareword
things\.  What about print vnnn "stuff\\n"\.  My point is\, even if 
you can make it "work\," what simple rule could there be that a 
programmer could use to decide what interpretation vnnn will have?
I don't think you can construct such a rule\.

It wouldn't surprise me that it doesn't work\.  However\, if it does
work\, it will certainly work/not work in surprising and confusing
ways\.  NO ONE will ever be able to explain exactly how/when this
bareword\-like thing works\.  Do you agree that that is unnecessary 
and bad?

I think an alternative syntax with braces fixes the probably rather
elegantly\.

\* >The current "smarts" that make v200\.200 bytes and v200\.2000 characters
\* >are Bad Bad Bad\.  I hope you can see this\.  What if I want v200\.200
\* >to be characters?
\*
\* Er\, v200\.200 is \_always\_ characters\.  What makes you think it isn't?

I'm trying to adopt Larry's line of thinking that it won't matter how
it's encoded\.  As long as that's true\, sure\, it makes sense to me\.

&nbsp;  \-joseph

**Comment by p5pRT** [comment]
### From @gisle

Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:

> On Thu\, 23 Mar 2000 23&#8203;:30&#8203;:48 MST\, "Joseph N\. Hall" wrote&#8203;:
> >Okay\, I am looking over these issues with 5\.6\.0 and I still see \.\.\.
> >
> >  "\\x\{c8\}" still equivalent to "\\xc8"&#8203;: 1 byte w/o utf8\, 2 bytes with
> 
> I think \\xc8 being one or two bytes was intentional \(to preserve
> compatibility\)\.  But "\\x\{NN\}" behaving the same way as "\\xNN"
> \(for 127 \< NN \< 256\) could be construed a misfeature\, because there's
> no compatibility issue to worry about with \\x\{\.\.\.\}\.

I disagree\.  It should not matter what we do\.

> >use utf8;
> >print "\\\\x\{41\} and v65 are different\\n" unless "\\x\{41\}" eq \+v65;
> >print "\\\\x41 and v65 are different\\n" unless "\\x41" eq \+v65;
> >print "\\\\x\{c8\} and v200 are different\\n" unless "\\x\{c8\}" eq \+v200;
> >print "\\\\xc8 and v200 are different\\n" unless "\\xc8" eq \+v200;
> >print "\\\\x\{221b\} and v8731 are different\\n" unless "\\x\{221b\}" eq v8731;
> >
> >still yields
> >
> >\\x\{c8\} and v200 are different
> >\\xc8 and v200 are different
> 
> This is the same issue as the first one\.

I think we need to patch sv\_eq similar to what I did for sv\_cmp at
some point\.  This is a patch&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- sv.c.dist	Fri Mar 24 22:37:54 2000
+++ sv.c	Fri Mar 24 22:44:46 2000
@@ -3920,10 +3920,19 @@
     else
 	pv1 = SvPV(str1, cur1);
 
-    if (!str2)
-	return !cur1;
-    else
-	pv2 = SvPV(str2, cur2);
+    if (cur1) {
+	if (!str2)
+	    return 0;
+	if (SvUTF8(str1) != SvUTF8(str2)) {
+	    if (SvUTF8(str1)) {
+		sv_utf8_upgrade(str2);
+	    }
+	    else {
+		sv_utf8_upgrade(str1);
+	    }
+	}
+    }
+    pv2 = SvPV(str2, cur2);
 
     if (cur1 != cur2)
 	return 0;



```
</details>
This breaks some bad tests in pragma/utf8 and lib/charnames.

Regards\,
Gisle

**Comment by p5pRT** [comment]
### From @TimToady

Gisle Aas writes&#8203;:
: Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
: 
: > On Thu\, 23 Mar 2000 23&#8203;:30&#8203;:48 MST\, "Joseph N\. Hall" wrote&#8203;:
: > >Okay\, I am looking over these issues with 5\.6\.0 and I still see \.\.\.
: > >
: > >  "\\x\{c8\}" still equivalent to "\\xc8"&#8203;: 1 byte w/o utf8\, 2 bytes with
: > 
: > I think \\xc8 being one or two bytes was intentional \(to preserve
: > compatibility\)\.  But "\\x\{NN\}" behaving the same way as "\\xNN"
: > \(for 127 \< NN \< 256\) could be construed a misfeature\, because there's
: > no compatibility issue to worry about with \\x\{\.\.\.\}\.
: 
: I disagree\.  It should not matter what we do\.

If I follow what you're saying\, I think I agree with Gisle\.  Whether
\\x takes brackets or not should in the long run be a purely syntactic
decision\.  Regardless of whether you write \\xc8 or \\x\{c8\}\, the decision
to encode into an 8\-bit string or a utf8 string should depend on context\,
not on whether you used the brackets\.

But by the same token\, it shouldn't matter whether they're encoded the
same as long as the context maintains the character abstraction\.
However\, I think perlunicode overstates when it says Perl will
eventually use characters \*instead\* of bytes\.  I think we'll always
have bytes\.  And there may well be situations in which it makes more
sense to map \\xc8 into 8\-bit representation than into utf8\.

: > >use utf8;
: > >print "\\\\x\{41\} and v65 are different\\n" unless "\\x\{41\}" eq \+v65;
: > >print "\\\\x41 and v65 are different\\n" unless "\\x41" eq \+v65;
: > >print "\\\\x\{c8\} and v200 are different\\n" unless "\\x\{c8\}" eq \+v200;
: > >print "\\\\xc8 and v200 are different\\n" unless "\\xc8" eq \+v200;
: > >print "\\\\x\{221b\} and v8731 are different\\n" unless "\\x\{221b\}" eq v8731;
: > >
: > >still yields
: > >
: > >\\x\{c8\} and v200 are different
: > >\\xc8 and v200 are different
: > 
: > This is the same issue as the first one\.
: 
: I think we need to patch sv\_eq similar to what I did for sv\_cmp at
: some point\.  This is a patch&#8203;:

Yes\, sv\_eq must have the same semantics as \!sv\_cmp\.

Larry

**Comment by p5pRT** [comment]
### From @TimToady

Larry Wall writes&#8203;:
: have bytes\.  And there may well be situations in which it makes more
: sense to map \\xc8 into 8\-bit representation than into utf8\.

Oops\, I meant \\x\{c8\}\.  But that just proves my point\.  They should mean
exactly the same thing\.

Larry
